File: SleepTimer/ViewModels/SleepTimerViewModel.swift
import Foundation
import Combine

class SleepTimerViewModel: ObservableObject {
    @Published var remainingTime: TimeInterval = 0
    @Published var isTimerRunning = false
    @Published var selectedSound: Sound?
    
    private var soundManager = SoundManager()
    private var timerCancellable: AnyCancellable?
    private var timer: Timer?
    
    func startTimer(minutes: Int) {
        guard !isTimerRunning else { return }
        
        remainingTime = TimeInterval(minutes * 60)
        isTimerRunning = true
        
        // Start playing the selected sound when timer starts
        if let sound = selectedSound {
            soundManager.playSound(sound)
        }
        
        timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            guard let self = self else { return }
            
            if self.remainingTime > 0 {
                self.remainingTime -= 1
            } else {
                self.stopTimer()
            }
        }
    }
    
    func stopTimer() {
        timer?.invalidate()
        timer = nil
        isTimerRunning = false
        soundManager.stopSound()
    }
    
    func selectSound(_ sound: Sound) {
        soundManager.stopSound()
        selectedSound = sound
        
        // Only play the sound if the timer is running
        if isTimerRunning {
            soundManager.playSound(sound)
        }
    }
    
    func formatTime(_ time: TimeInterval) -> String {
        let hours = Int(time) / 3600
        let minutes = Int(time) % 3600 / 60
        let seconds = Int(time) % 60
        
        if hours > 0 {
            return String(format: "%d:%02d:%02d", hours, minutes, seconds)
        } else {
            return String(format: "%02d:%02d", minutes, seconds)
        }
    }
    
    deinit {
        stopTimer()
    }
} 

File: app/src/main/java/com/example/sleeptimer/ui/SleepTimerViewModel.kt
package com.example.sleeptimer.ui

import android.content.Context
import android.media.MediaPlayer
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class SleepTimerViewModel : ViewModel() {
    var remainingTime by mutableStateOf(0L)
        private set
    
    var isTimerRunning by mutableStateOf(false)
        private set
    
    var selectedSound by mutableStateOf<Sound?>(null)
        private set
    
    private var timerJob: Job? = null
    private var mediaPlayer: MediaPlayer? = null
    
    fun startTimer(minutes: Int, context: Context) {
        if (isTimerRunning) return
        
        remainingTime = minutes * 60L
        isTimerRunning = true
        
        // Start playing the selected sound when timer starts
        selectedSound?.let { sound ->
            playSound(sound, context)
        }
        
        timerJob = viewModelScope.launch {
            while (remainingTime > 0) {
                delay(1000)
                remainingTime--
            }
            // Timer has ended
            stopTimer()
            stopSound() // Ensure sound is stopped when timer ends
        }
    }
    
    fun stopTimer() {
        timerJob?.cancel()
        timerJob = null
        isTimerRunning = false
        stopSound()
    }
    
    fun selectSound(sound: Sound, context: Context) {
        stopSound()
        selectedSound = sound
        // Only play the sound if the timer is running
        if (isTimerRunning) {
            playSound(sound, context)
        }
    }
    
    private fun playSound(sound: Sound, context: Context) {
        stopSound()
        mediaPlayer = MediaPlayer.create(context, sound.resourceId).apply {
            isLooping = true
            start()
        }
    }
    
    private fun stopSound() {
        mediaPlayer?.apply {
            if (isPlaying) {
                stop()
            }
            release()
        }
        mediaPlayer = null
    }
    
    override fun onCleared() {
        super.onCleared()
        stopTimer()
        stopSound()
    }
} 